http/digest/src/main/java/org/wildfly/security/http/digest/PersistentNonceManager.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2023 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.http.digest;
+
+import java.io.Serializable;
+import java.security.SecureRandom;
+import java.util.HashMap;
+import java.util.Map;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.wildfly.security.mechanism.AuthenticationMechanismException;
+import org.wildfly.security.mechanism._private.ElytronMessages;
+
+/**
+ * A utility responsible for managing nonces that can be stored in an HTTP session.
+ */
+public class PersistentNonceManager extends NonceManager implements Serializable {
+
+    private transient ScheduledExecutorService executor;
+    private AtomicInteger nonceCounter = new AtomicInteger();
+    private Map<String, NonceManagerUtils.NonceState> usedNonces = new HashMap<>();
+    private byte[] privateKey;
+    private long validityPeriodNano;
+    private long nonceSessionTime;
+    private boolean singleUse;
+    private String algorithm;
+    private volatile ElytronMessages log;
```

Comment:
Why is this volatile?

Reply:
@fjuma Not sure why this was volatile, I removed it, thanks!